tests/test_directives.py

```diff
@@ -123,45 +124,90 @@ class TestDirectives:
+        directives._set_defined_directive(NP.name, 10)
+        assert directives[NP.name] == 10
 
-    def test_define_directive(self, setUp, available_directives_list):
-        directives = setUp(available_directives_list=available_directives_list[:-1])
-        # The values below corresponds to NP, NRANKS, NGPU, EXECUTABLE.
-        valid_values = [100, 100, 100, 'My String']
-        invalid_values = [0, -1, -1, [0]]
-        for idx, dir in enumerate(available_directives_list[:-1]):
-            directives._set_defined_directive(dir.name, valid_values[idx])
-            assert directives[dir.name] == valid_values[idx]
-        for idx, dir in enumerate(available_directives_list[:-1]):
-            with pytest.raises(ValueError):
-                directives._set_defined_directive(dir.name, invalid_values[idx])
-
-    def test_del_directive(self, setUp, available_directives_list):
-        directives = setUp(available_directives_list=available_directives_list[:-1])
-        directives[OMP_NUM_THREADS.name] = OMP_NUM_THREADS.default
+    def test_set_defined_directive_invalid(self, setUp, available_directives_list):
+        directives = setUp(available_directives_list=available_directives_list)
+        directives._set_defined_directive(NP.name, 0)
+        assert directives[NP.name] == NP.default
+
+    def test_set_undefined_directive(self, setUp, available_directives_list, product_directive):
+        directives = setUp(available_directives_list=available_directives_list)
+        with pytest.raises(Exception):
+            directives._set_defined_directive(product_directive.name, 0)
+
+    def test_set_directives_item(self, setUp, available_directives_list, product_directive):
+        directives = setUp(available_directives_list=available_directives_list)
+        directives[product_directive.name] = product_directive.default
+        assert directives[product_directive.name] == product_directive.default
+
+    def test_del_directive(self, setUp, available_directives_list, product_directive):
+        directives = setUp(available_directives_list=available_directives_list)
+        directives[product_directive.name] = 100
         directives._set_defined_directive(NP.name, 100)
         assert directives[NP.name] == 100
-        assert directives[OMP_NUM_THREADS.name] == OMP_NUM_THREADS.default
+        assert directives[product_directive.name] == 100
         del directives[NP.name]
         assert directives[NP.name] == NP.default
-        del directives[OMP_NUM_THREADS.name]
+        del directives[product_directive.name]
         with pytest.raises(KeyError):
-            directives[OMP_NUM_THREADS.name]
+            directives[product_directive.name]
+
+    def test_update_directive_without_aggregate(self, setUp, available_directives_list):
+        directives1 = setUp(available_directives_list=available_directives_list)
+        directives2 = deepcopy(directives1)
```

Comment:
You don't need another `Directives` class here. You could just use `directives1.update(valid_values)` and `assert directives1[key] == valid_values[key]`.

Reply:
An unexpected error occurs here. 
While updating I set 
```
valid_values = {'np': 4, 'ngpu': 1, 'nranks': 0,
                        'omp_num_threads': 10, 'executable': 'PathFinder',
                        'walltime': 20., 'memory': 16, 'processor_fraction': 0.5}
```
The `np` directive is updated to 4 while it updates every other directive properly. I checked that this functionality arrises due to the finalize attribute of `NP`. Am I correct? If yes then how did it work when I used another `Directives` class

Reply:
Also after digging into the code, I realized that when we set any directive then it calls the directive. For instance, if we set the value for NP = 4, then it'll eventually result in this statement NP(4) which, after validation, returns some value and that value is set to that particular directive in the `Directives` class. I check the validation function of NP, the post-process here is the `raise_if_too_small` method. 

I am not able to see any involvement of the finalize attribute here. Can you please help me?